### DISCUSSION TITLE
Why there is additional spaces after each character when trying to get textline in pdf?

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#3469] [General]
[p2.pdf](https://github.com/pymupdf/PyMuPDF/files/15289427/p2.pdf)
'''
    doc = fitz.open(path)

    j = 0
    k = 0
    s3 = set()
    for i in range(0, doc.page_count):
        print(f'开始识别页码：{str(i + 1)}/{doc.page_count}')
        k = k + 1
        page = doc[i]
        for block in page.get_text("dict", flags=fitz.TEXTFLAGS_TEXT)["blocks"]:
            for line in block["lines"]:
                j = j + 1
                bbox = line["bbox"]  # the line bbox
                text = ''.join([span["text"] for span in line["spans"]])  # text in line

                # print(bbox)
                print(text)
'''

line result example is '【 請 求 項 １ 】'
When copying directly with Adobe, result is "【請求項１ 】", no additional spaces

**Comment by nissansz** [comment]
[cnp1.pdf](https://github.com/pymupdf/PyMuPDF/files/15290594/cnp1.pdf)
Another problem is, the bbox is smaller than real text, and the cropped image may cut the final character to half.

Line 2 in attached pdf is wrongly cropped, and result is as follows.
![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/1d26738b-6f40-4a93-92cf-0a0c2f704679)

**Comment by JorjMcKie** [comment]
This is largely a problem introduced by the standard text flags. Things will improve if you only use `fitz.TEXT_DEHYPHENATE | fitz.TEXT_MEDIABOX_CLIP`.

Otherwise, you may be interested in looking at the new package / repository [pymupdf4llm](https://pypi.org/project/pymupdf4llm/).
It will produce markdown text output, but also detects multiple text columns and tables and creates a unified output across all of this content. When using it with your 2 examples like this:
```python
import pathlib
import pymupdf4llm

data=pymupdf4llm.to_markdown("cnp1.pdf")
pathlib.Path("cnp1.pdf.md").write_bytes(data.encode())

data=pymupdf4llm.to_markdown("p2.pdf")
pathlib.Path("p2.pdf.md").write_bytes(data.encode())
```
You will receive these output files: 
[cnp1.pdf.md](https://github.com/pymupdf/PyMuPDF/files/15389511/cnp1.pdf.md)
[p2.pdf.md](https://github.com/pymupdf/PyMuPDF/files/15389512/p2.pdf.md)

**Reply by nissansz** [reply]
How to use  fitz.TEXT_DEHYPHENATE | fitz.TEXT_MEDIABOX_CLIP to parse each text line?

**Reply by JorjMcKie** [reply]
set text extraction flags to this value!

**Reply by nissansz** [reply]
How to revise in below code?

doc = fitz.open(path)

j = 0
k = 0
s3 = set()
for i in range(0, doc.page_count):
    print(f'开始识别页码：{str(i + 1)}/{doc.page_count}')
    k = k + 1
    page = doc[i]
    for block in page.get_text("dict", flags=fitz.TEXTFLAGS_TEXT)["blocks"]:
        for line in block["lines"]:
            j = j + 1
            bbox = line["bbox"]  # the line bbox
            text = ''.join([span["text"] for span in line["spans"]])  # text in line

            # print(bbox)
            print(text)

**Reply by JorjMcKie** [reply]
Instead of `flags=fitz.TEXTFLAGS_TEXT` use `flags=fitz.TEXT_DEHYPHENATE | fitz.TEXT_MEDIABOX_CLIP`

**Reply by nissansz** [reply]
Same wrong result.
